File: app/src/main/java/com/fatbook/fatbookapp/ui/fragment/UserProfileFragment.java
package com.fatbook.fatbookapp.ui.fragment;

import android.Manifest;
import android.app.AlertDialog;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.graphics.PorterDuff;
import android.net.Uri;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.WindowManager;
import android.widget.FrameLayout;
import android.widget.TextView;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.fragment.NavHostFragment;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.fatbook.fatbookapp.R;
import com.fatbook.fatbookapp.core.Recipe;
import com.fatbook.fatbookapp.core.User;
import com.fatbook.fatbookapp.databinding.FragmentUserProfileBinding;
import com.fatbook.fatbookapp.retrofit.RetrofitFactory;
import com.fatbook.fatbookapp.ui.activity.SplashActivity;
import com.fatbook.fatbookapp.ui.adapters.RecipeAdapter;
import com.fatbook.fatbookapp.ui.listeners.OnRecipeClickListener;
import com.fatbook.fatbookapp.ui.viewmodel.RecipeViewModel;
import com.fatbook.fatbookapp.ui.viewmodel.UserViewModel;
import com.fatbook.fatbookapp.util.FileUtils;
import com.fatbook.fatbookapp.util.KeyboardActionUtil;
import com.fatbook.fatbookapp.util.UserUtils;
import com.google.android.material.appbar.AppBarLayout;
import com.google.android.material.snackbar.Snackbar;

import org.apache.commons.lang3.StringUtils;

import java.io.File;
import java.util.logging.Level;

import lombok.extern.java.Log;
import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.RequestBody;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

@Log
public class UserProfileFragment extends Fragment implements OnRecipeClickListener {

    private FragmentUserProfileBinding binding;

    private User user;

    private UserViewModel userViewModel;

    private RecipeAdapter adapter;

    private ActivityResultLauncher<String> choosePhotoFromGallery;

    private File userPhoto;

    private Uri selectedImageUri;

    private static final int REQUEST_EXTERNAL_STORAGE = 1;
    private static final String[] PERMISSIONS_STORAGE = {
            Manifest.permission.READ_EXTERNAL_STORAGE,
            Manifest.permission.WRITE_EXTERNAL_STORAGE
    };

    private RecipeViewModel recipeViewModel;

    @Override
    public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        requireActivity().getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_RESIZE);

        ((AppBarLayout.LayoutParams) binding.toolbarUserProfile.getLayoutParams()).setScrollFlags(0);

        userViewModel = new ViewModelProvider(requireActivity()).get(UserViewModel.class);
        recipeViewModel = new ViewModelProvider(requireActivity()).get(RecipeViewModel.class);
        setupMenu();

        user = userViewModel.getUser().getValue();
        if (userViewModel.getUser().getValue() == null) {
            loadUser();
        }
        userViewModel.getUser().observe(getViewLifecycleOwner(), _user -> {
            binding.swipeRefreshUserProfile.setRefreshing(false);
            user = _user;
            fillUserProfile();
        });
        setupAdapter();
        setupSwipeRefresh();
        binding.toolbarUserProfile.getOverflowIcon().setColorFilter(getResources().getColor(R.color.color_blue_grey_600), PorterDuff.Mode.MULTIPLY);

        try {
            choosePhotoFromGallery = registerForActivityResult(new ActivityResultContracts.GetContent(), uri -> {
                verifyStoragePermissions(requireActivity());
                if (uri != null) {
                    selectedImageUri = uri;
                    String path = FileUtils.getPath(requireContext(), selectedImageUri);
                    userPhoto = new File(path);
                    binding.imageViewProfilePhoto.setImageURI(selectedImageUri);
                    binding.imageViewUserProfilePhotoBgr.setImageURI(selectedImageUri);

                    binding.buttonUserProfileAddPhoto.setVisibility(View.GONE);
                    binding.buttonUserProfileChangePhoto.setVisibility(View.VISIBLE);
                    binding.buttonUserProfileDeletePhoto.setVisibility(View.VISIBLE);
                } else {
                    binding.buttonUserProfileAddPhoto.setVisibility(View.VISIBLE);
                    binding.buttonUserProfileChangePhoto.setVisibility(View.GONE);
                    binding.buttonUserProfileDeletePhoto.setVisibility(View.GONE);
                }
            });
        } catch (Exception e) {
            e.printStackTrace();
        }

        binding.buttonUserProfileAddPhoto.setOnClickListener(view1 -> {
            verifyStoragePermissions(requireActivity());
            choosePhotoFromGallery.launch("image/*");
        });

        binding.buttonUserProfileChangePhoto.setOnClickListener(view1 -> {
            verifyStoragePermissions(requireActivity());
            choosePhotoFromGallery.launch("image/*");
        });

        binding.buttonUserProfileDeletePhoto.setOnClickListener(view1 -> {
            selectedImageUri = null;
            userPhoto = null;
            user.setImage(StringUtils.EMPTY);
            fillUserProfile();

            binding.buttonUserProfileAddPhoto.setVisibility(View.VISIBLE);
            binding.buttonUserProfileChangePhoto.setVisibility(View.GONE);
            binding.buttonUserProfileDeletePhoto.setVisibility(View.GONE);
        });
    }

    private void verifyStoragePermissions(FragmentActivity requireActivity) {
        int permission = ActivityCompat.checkSelfPermission(requireActivity, Manifest.permission.WRITE_EXTERNAL_STORAGE);

        if (permission != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(
                    requireActivity,
                    PERMISSIONS_STORAGE,
                    REQUEST_EXTERNAL_STORAGE
            );
        }
    }

    private void loadUser() {
        RetrofitFactory.apiServiceClient().getUser(user.getLogin()).enqueue(new Callback<User>() {
            @Override
            public void onResponse(@NonNull Call<User> call, @NonNull Response<User> response) {
                log.log(Level.INFO, "" + response.code() + " found user: " + response.body());
                userViewModel.setUser(response.body());
            }

            @Override
            public void onFailure(@NonNull Call<User> call, @NonNull Throwable t) {
                log.log(Level.INFO, "load user ERROR");
            }
        });
    }

    private void setupSwipeRefresh() {
        editMode(false);
        binding.swipeRefreshUserProfile.setColorSchemeColors(getResources().getColor(R.color.color_pink_a200));
        binding.swipeRefreshUserProfile.setOnRefreshListener(this::loadUser);
    }

    private void setupAdapter() {
        RecyclerView recyclerView = binding.rvUserRecipe;
        adapter = new RecipeAdapter(binding.getRoot().getContext(), user.getRecipes(), user, this);
        recyclerView.setAdapter(adapter);
    }

    private void fillUserProfile() {
        binding.toolbarUserProfile.setTitle(user.getLogin());
        binding.editTextProfileName.setText(user.getName());
        binding.editTextProfileBio.setText(user.getBio());
        if (StringUtils.isNotEmpty(user.getImage())) {
            Glide
                    .with(getLayoutInflater()
                            .getContext())
                    .load(user.getImage())
                    .into(binding.imageViewProfilePhoto);

            Glide
                    .with(getLayoutInflater()
                            .getContext())
                    .load(user.getImage())
                    .into(binding.imageViewUserProfilePhotoBgr);
        } else {
            binding.imageViewProfilePhoto.setImageDrawable(getResources().getDrawable(R.drawable.image_recipe_default));
            binding.imageViewUserProfilePhotoBgr.setImageDrawable(getResources().getDrawable(R.drawable.user_profile_default_bgr));
        }
        adapter.setData(user.getRecipes(), user);
        adapter.notifyDataSetChanged();
    }

    private void setupMenu() {
        AppCompatActivity activity = (AppCompatActivity) getActivity();
        assert activity != null;
        activity.setSupportActionBar(binding.toolbarUserProfile);
        setHasOptionsMenu(true);
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        inflater.inflate(R.menu.user_profile_menu, menu);
        super.onCreateOptionsMenu(menu, inflater);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_user_profile_edit:
                editMode(true);
                return true;
            case R.id.menu_user_profile_save:
                editMode(false);
                confirmEdit();
                return true;
            case R.id.menu_user_profile_cancel:
                editMode(false);
                revertEdit();
                return true;
            case R.id.menu_user_profile_app_info:
                Snackbar.make(binding.getRoot(), "TODO app info", Snackbar.LENGTH_SHORT).setAnchorView(requireActivity().findViewById(R.id.bottom_navigation)).show();
                return true;
            case R.id.menu_user_profile_logout:
                logout();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private void logout() {
        String msg = getResources().getString(R.string.alert_dialog_logout_message);

        final TextView textViewMsg = new TextView(requireContext());
        textViewMsg.setText(msg);
        textViewMsg.setSingleLine();
        textViewMsg.setTextColor(getResources().getColor(R.color.color_blue_grey_600));

        FrameLayout container = new FrameLayout(requireContext());
        FrameLayout.LayoutParams params = new FrameLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        params.leftMargin = getResources().getDimensionPixelSize(R.dimen.dialog_margin);
        params.rightMargin = getResources().getDimensionPixelSize(R.dimen.dialog_margin);
        params.topMargin = getResources().getDimensionPixelSize(R.dimen.dialog_margin);
        textViewMsg.setLayoutParams(params);
        container.addView(textViewMsg);

        View title = LayoutInflater.from(requireContext()).inflate(R.layout.alert_dialog_title_logout, null);
        new AlertDialog.Builder(requireContext()).setView(container).setCustomTitle(title).setPositiveButton(getString(R.string.alert_dialog_btn_yes), (dialogInterface, i) -> {
            SharedPreferences sharedPreferences = requireActivity().getSharedPreferences(UserUtils.APP_PREFS, Context.MODE_PRIVATE);
            SharedPreferences.Editor editor = sharedPreferences.edit();
            editor.putString(UserUtils.USER_LOGIN, StringUtils.EMPTY);
            editor.apply();
            startActivity(new Intent(requireActivity(), SplashActivity.class));
            requireActivity().finish();
        }).setNegativeButton(getString(R.string.alert_dialog_btn_cancel), (dialogInterface, i) -> {
            dialogInterface.dismiss();
        }).show();
    }

    private void editMode(boolean allow) {
        changeMenuItemsVisibility(!allow, allow, allow);

        binding.editTextProfileName.setEnabled(allow);
        binding.editTextProfileBio.setEnabled(allow);
        if (allow) {
            binding.linearlayoutButtonsUserImage.setVisibility(View.VISIBLE);
            if (userPhoto == null && StringUtils.isEmpty(user.getImage())) {
                binding.buttonUserProfileAddPhoto.setVisibility(View.VISIBLE);
                binding.buttonUserProfileChangePhoto.setVisibility(View.GONE);
                binding.buttonUserProfileDeletePhoto.setVisibility(View.GONE);
            } else {
                binding.buttonUserProfileAddPhoto.setVisibility(View.GONE);
                binding.buttonUserProfileChangePhoto.setVisibility(View.VISIBLE);
                binding.buttonUserProfileDeletePhoto.setVisibility(View.VISIBLE);
            }
            binding.imageViewUserProfilePhotoBgr.setVisibility(View.INVISIBLE);
            binding.editTextProfileName.setBackgroundResource(R.drawable.edit_mode_bgr);
            binding.editTextProfileBio.setBackgroundResource(R.drawable.edit_mode_bgr);

        } else {
            binding.linearlayoutButtonsUserImage.setVisibility(View.GONE);
            binding.imageViewUserProfilePhotoBgr.setVisibility(View.VISIBLE);
            binding.editTextProfileName.setBackgroundResource(R.drawable.round_corner_rect_white);
            binding.editTextProfileBio.setBackgroundResource(R.drawable.round_corner_rect_white);
        }
    }

    private void revertEdit() {
        fillUserProfile();
    }

    private void confirmEdit() {
        String strName = binding.editTextProfileName.getText().toString();
        binding.editTextProfileName.setText(strName.trim());

        String strBio = binding.editTextProfileBio.getText().toString();
        binding.editTextProfileBio.setText(strBio.replace("\n", " ").trim());

        user.setName(binding.editTextProfileName.getText().toString());
        user.setBio(binding.editTextProfileBio.getText().toString());

        saveUser();
    }

    private void saveUser() {
        if (userPhoto != null) {
            try {
                RequestBody requestFile = RequestBody.create(MediaType.parse("image/*"), userPhoto);
                String fileName = "image" + userPhoto.getName().substring(userPhoto.getName().indexOf('.'));
                MultipartBody.Part file = MultipartBody.Part.createFormData("file", fileName, requestFile);

                RetrofitFactory.apiServiceClient().uploadUserImage(file, FileUtils.TAG_USER, user.getPid()).enqueue(new Callback<User>() {
                    @Override
                    public void onResponse(@NonNull Call<User> call, @NonNull Response<User> response) {
                        if (response.code() == 200) {
                            log.log(Level.INFO, "image add SUCCESS");
                            userViewModel.setUser(response.body());
                        } else {
                            log.log(Level.INFO, "image add FAILED " + response.code());
                        }
                    }

                    @Override
                    public void onFailure(@NonNull Call<User> call, @NonNull Throwable t) {
                        log.log(Level.INFO, "image add FAILED");
                    }
                });
            } catch (Exception e) {
                log.log(Level.INFO, e.toString());
                e.printStackTrace();
            }
        }
    }

    private void changeMenuItemsVisibility(boolean edit, boolean save, boolean cancel) {
        if (binding.toolbarUserProfile.getMenu().findItem(R.id.menu_user_profile_edit) != null) {
            binding.toolbarUserProfile.getMenu().findItem(R.id.menu_user_profile_edit).setVisible(edit);
            binding.toolbarUserProfile.getMenu().findItem(R.id.menu_user_profile_save).setVisible(save);
            binding.toolbarUserProfile.getMenu().findItem(R.id.menu_user_profile_cancel).setVisible(cancel);
        }
    }

    @Override
    public void onRecipeClick(int position) {
        Recipe recipe = user.getRecipes().get(position);
        recipeViewModel.setSelectedRecipe(recipe);
        recipeViewModel.setSelectedRecipePosition(position);
        NavHostFragment.findNavController(this).navigate(R.id.action_go_to_recipe_view_from_user_profile);
    }

    @Override
    public void onBookmarksClick(Recipe recipe, boolean add, int adapterPosition) {
        System.out.println("Stub!");
    }

    @Override
    public void onForkClicked(Recipe recipe, boolean fork, int position) {
        recipeForked(recipe, fork, position);
    }

    private void recipeForked(Recipe recipe, boolean fork, int position) {
        RetrofitFactory.apiServiceClient().recipeForked(user.getPid(), recipe.getPid(), fork).enqueue(new Callback<Recipe>() {
            @Override
            public void onResponse(@NonNull Call<Recipe> call, @NonNull Response<Recipe> response) {
                log.log(Level.INFO, "fork SUCCESS");
                recipeViewModel.setSelectedRecipe(response.body());
                loadUser();
            }

            @Override
            public void onFailure(@NonNull Call<Recipe> call, @NonNull Throwable t) {
                log.log(Level.INFO, "fork FAILED");
            }
        });
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = FragmentUserProfileBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onResume() {
        super.onResume();
        binding.getRoot().getViewTreeObserver().addOnGlobalLayoutListener(new KeyboardActionUtil(binding.getRoot(), requireActivity()).listenerForAdjustResize);
        if (selectedImageUri != null) {
            binding.imageViewProfilePhoto.setImageURI(selectedImageUri);
            binding.imageViewUserProfilePhotoBgr.setImageURI(selectedImageUri);
        }
    }

    @Override
    public void onPause() {
        super.onPause();
        binding.getRoot().getViewTreeObserver().removeOnGlobalLayoutListener(new KeyboardActionUtil(binding.getRoot(), requireActivity()).listenerForAdjustResize);
    }

//    private void setupKeyboardOpenListener() {
//        keyboardVisibilityListener = () -> {
//            final Rect rectangle = new Rect();
//            final View contentView = binding.getRoot();
//            contentView.getWindowVisibleDisplayFrame(rectangle);
//            int screenHeight = contentView.getRootView().getHeight();
//            int keypadHeight = screenHeight - rectangle.bottom;
//            boolean isKeyboardNowVisible = keypadHeight > screenHeight * 0.15;
//
//            if (isKeyboardVisible != isKeyboardNowVisible) {
//                if (isKeyboardNowVisible) {
//                    requireActivity().findViewById(R.id.bottom_navigation).setVisibility(View.GONE);
//                } else {
//                    requireActivity().findViewById(R.id.bottom_navigation).setVisibility(View.VISIBLE);
//                }
//            }
//            isKeyboardVisible = isKeyboardNowVisible;
//        };
//    }

}

File: app/src/main/java/online/fatbook/fatbookapp/ui/fragment/user/UserProfileFragment.kt
package online.fatbook.fatbookapp.ui.fragment.user

import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.*
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.core.widget.NestedScrollView
import androidx.fragment.app.Fragment
import androidx.navigation.fragment.NavHostFragment
import androidx.transition.AutoTransition
import androidx.transition.Scene
import androidx.transition.TransitionManager
import androidx.viewpager2.adapter.FragmentStateAdapter
import androidx.viewpager2.widget.ViewPager2
import com.bumptech.glide.Glide
import com.google.android.material.tabs.TabLayoutMediator
import kotlinx.android.synthetic.main.fragment_user_profile.*
import kotlinx.android.synthetic.main.fragment_user_profile.nsv_userprofile
import kotlinx.android.synthetic.main.include_progress_overlay.*
import online.fatbook.fatbookapp.R
import online.fatbook.fatbookapp.callback.ResultCallback
import online.fatbook.fatbookapp.core.user.User
import online.fatbook.fatbookapp.databinding.FragmentUserProfileBinding
import online.fatbook.fatbookapp.ui.activity.SplashActivity
import online.fatbook.fatbookapp.ui.adapters.UserProfileRecipesAdapter
import online.fatbook.fatbookapp.ui.listeners.BaseFragmentActionsListener
import online.fatbook.fatbookapp.ui.viewmodel.ImageViewModel
import online.fatbook.fatbookapp.ui.viewmodel.UserViewModel
import online.fatbook.fatbookapp.util.*
import online.fatbook.fatbookapp.util.alert_dialog.FBAlertDialogBuilder
import online.fatbook.fatbookapp.util.alert_dialog.FBAlertDialogListener
import org.apache.commons.lang3.StringUtils


class UserProfileFragment : Fragment(), BaseFragmentActionsListener {

    private var binding: FragmentUserProfileBinding? = null
    private val userViewModel by lazy { obtainViewModel(UserViewModel::class.java) }
    private val imageViewModel by lazy { obtainViewModel(ImageViewModel::class.java) }

    private var expanded: Boolean = false
    private lateinit var viewPager: ViewPager2
    private lateinit var fragmentAdapter: UserProfileRecipesAdapter

    companion object {
        private const val TAG = "UserProfileFragment"
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?
    ): View {
        binding = FragmentUserProfileBinding.inflate(inflater, container, false)
        return binding!!.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        Log.d("UserProfileFragment", "onViewCreated")
        val sharedPreferences =
            requireActivity().getSharedPreferences(Constants.SP_TAG, AppCompatActivity.MODE_PRIVATE)
        if (sharedPreferences.getBoolean(Constants.SP_TAG_DARK_MODE_CHANGED, false)) {
            sharedPreferences.edit().putBoolean(Constants.SP_TAG_DARK_MODE_CHANGED, false).apply()
            NavHostFragment.findNavController(this)
                .navigate(R.id.action_go_to_app_settings_from_user_profile)
        } else {
            progress_overlay.visibility = View.VISIBLE
            toolbar_userprofile.visibility = View.GONE
            loadUserData()
            setupMenu(R.menu.user_profile_current_menu)
            setupSwipeRefresh()
            imageview_recipes_qtt_userprofile.setOnClickListener {
                focusOnRecipes()
            }

            imageview_friends_qtt_userprofile.setOnClickListener {
                NavHostFragment.findNavController(this)
                    .navigate(R.id.action_go_to_followers_from_user_profile)
            }

            imageview_userphoto_userprofile.setOnClickListener {
                imageViewModel.image.value = userViewModel.user.value!!.profileImage
                NavHostFragment.findNavController(this)
                    .navigate(R.id.action_go_to_view_image_from_user_profile1)
            }

            nsv_userprofile.setOnScrollChangeListener(NestedScrollView.OnScrollChangeListener { _, _, scrollY, _, _ ->
                showButtonUp(scrollY)
            })

            floating_button_up.setOnClickListener {
                focusOnRecipes()
            }

            imageview_ic_expand.setOnClickListener {
                animateTextExpand()
            }
        }
    }

    private fun setupSwipeRefresh() {
        swipe_refresh_user_profile.isEnabled = false
        swipe_refresh_user_profile.isRefreshing = false
        swipe_refresh_user_profile.setProgressBackgroundColorSchemeColor(
            ContextCompat.getColor(
                requireContext(),
                R.color.theme_primary_bgr
            )
        )
        swipe_refresh_user_profile.setColorSchemeColors(
            ContextCompat.getColor(
                requireContext(), R.color.color_pink_a200
            )
        )
        swipe_refresh_user_profile.setOnRefreshListener {
            loadUserData()
        }
    }

    private fun loadUserData() {
        if (userViewModel.selectedUsername.value.isNullOrEmpty()) {
            toolbar_userprofile.title = userViewModel.user.value!!.username
            setupViewForLoggedInUser()
        } else {
            toolbar_userprofile.title = userViewModel.selectedUsername.value!!
            setupViewForSelectedUser()
        }
    }

    private fun setupMenu(menu: Int) {
        toolbar_userprofile.inflateMenu(menu)
        toolbar_userprofile.setOnMenuItemClickListener(this::onOptionsItemSelected)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.menu_user_profile_edit_profile -> {
                NavHostFragment.findNavController(this)
                    .navigate(R.id.action_go_to_edit_profile_from_user_profile)
                true
            }
            R.id.menu_user_profile_badges -> {
                NavHostFragment.findNavController(this)
                    .navigate(R.id.action_go_to_badges_from_user_profile)
                true
            }
            R.id.menu_user_profile_app_settings -> {
                NavHostFragment.findNavController(this)
                    .navigate(R.id.action_go_to_app_settings_from_user_profile)
                true
            }
            R.id.menu_user_profile_app_info -> {
                showAppInfoDialog()
                true
            }
            R.id.menu_user_profile_logout -> {
                logout()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun showAppInfoDialog() {
        FBAlertDialogBuilder.getAppInfoDialog(object : FBAlertDialogListener {
            override fun onClick(dialogInterface: DialogInterface) {
                dialogInterface.dismiss()
            }
        }).show()
    }

    private fun logout() {
        val sharedPreferences = requireActivity().getSharedPreferences(
            Constants.SP_TAG, Context.MODE_PRIVATE
        )
        val editor = sharedPreferences.edit()
        editor.putString(Constants.SP_TAG_USERNAME, StringUtils.EMPTY)
        editor.putString(Constants.SP_TAG_PASSWORD, StringUtils.EMPTY)
        editor.apply()
        startActivity(Intent(requireActivity(), SplashActivity::class.java))
        requireActivity().finish()
    }

    private fun animateTextExpand() {
        TransitionManager.go(Scene(cardview_userprofile), AutoTransition())
        if (!expanded) {
            textview_bio_userprofile.maxLines = Integer.MAX_VALUE
            imageview_ic_expand.setImageDrawable(
                ContextCompat.getDrawable(
                    requireContext(), R.drawable.ic_expand_less
                )
            )
            expanded = true
        } else {
            textview_bio_userprofile.maxLines = 3
            imageview_ic_expand.setImageDrawable(
                ContextCompat.getDrawable(
                    requireContext(), R.drawable.ic_expand_more
                )
            )
            expanded = false
        }
    }

    private fun showButtonUp(scrollY: Int) {
        if (!expanded) {
            if (scrollY >= 1044) {
                floating_button_up.visibility = View.VISIBLE
            }
            if (scrollY < 1044) {
                floating_button_up.visibility = View.GONE
            }
        } else {
            if (scrollY >= 1200) {
                floating_button_up.visibility = View.VISIBLE
            }
            if (scrollY < 1200) {
                floating_button_up.visibility = View.GONE
            }
        }
    }

    private fun setupViewForLoggedInUser() {
        ll_btns_follow_message.visibility = View.GONE
        tabLayout_userprofile.visibility = View.VISIBLE
        toolbar_userprofile.navigationIcon = null
//        if (userViewModel.user.value?.pid == null) {
        loadUser(userViewModel.user.value!!.username!!, true)
//        } else {
//            drawData(userViewModel.user.value!!)
//        }
    }

    private fun setupViewForSelectedUser() {
        ll_btns_follow_message.visibility = View.VISIBLE
        tabLayout_userprofile.visibility = View.GONE
        toolbar_userprofile.navigationIcon = context?.getDrawable(R.drawable.ic_arrow_back)
        if (userViewModel.selectedUser.value?.pid == null) {
            loadUser(userViewModel.selectedUsername.value!!, false)
        } else {
            drawData(userViewModel.selectedUser.value!!)
        }
    }

    private fun drawData(user: User) {
        Log.i("DRAW DATA", "--------------------------------------------------")
        toolbar_userprofile.title = user.username
        if (user.online) {
            imageview_is_online.visibility = View.VISIBLE
            toolbar_userprofile.subtitle = getString(R.string.subtitle_online)
        } else {
            imageview_is_online.visibility = View.INVISIBLE
            toolbar_userprofile.subtitle = getString(R.string.subtitle_offline)
        }
        textview_recipes_qtt_userprofile.text =
            user.recipeAmount?.let { FormatUtils.prettyCount(it) }
        textview_friends_qtt_userprofile.text =
            user.followersAmount?.let { FormatUtils.prettyCount(it) }
        if (user.profileImage.isNullOrEmpty()) {
            imageview_userphoto_userprofile.setImageDrawable(requireContext().getDrawable(R.drawable.ic_default_userphoto))
            imageview_userphoto_userprofile.isClickable = false
        } else {
            Glide.with(requireContext()).load(user.profileImage!!)
                .into(imageview_userphoto_userprofile)
            imageview_userphoto_userprofile.isClickable = true
        }
        if (user.title.isNullOrEmpty()) {
            textview_title_userprofile.visibility = View.GONE
        } else {
            textview_title_userprofile.text = user.title
        }

        if (user.website.isNullOrEmpty()) {
            textview_website_userprofile.visibility = View.GONE
            imageview_ic_website.visibility = View.GONE
        } else {
            textview_website_userprofile.text = user.website
        }

        if (user.bio.isNullOrEmpty()) {
            expandableLayout.visibility = View.GONE
            imageview_ic_expand.visibility = View.GONE
        } else {
            textview_bio_userprofile.text = user.bio
            if (textview_bio_userprofile.lineCount <= textview_bio_userprofile.maxLines) {
                imageview_ic_expand.visibility = View.GONE
            } else {
                imageview_ic_expand.visibility = View.VISIBLE
            }
        }
        progress_overlay.visibility = View.GONE
        toolbar_userprofile.visibility = View.VISIBLE
        swipe_refresh_user_profile.isRefreshing = false
        swipe_refresh_user_profile.isEnabled = true

//        userViewModel.user.value!!.recipesFavourites?.let {
//            if (it.isNotEmpty()) {
//                Log.d(TAG, "favs: ${userViewModel.user.value!!.recipesFavourites}")
////                fragmentAdapter.setFavs()
//            }
//        }
//        userViewModel.user.value!!.recipes?.let {
//            if (it.isNotEmpty()) {
//                Log.d(TAG, "recipies: ${userViewModel.user.value!!.recipes}")
////                fragmentAdapter.setRecipes()
//            }
//        }

        userViewModel.user.value.let {
            setupViewPager()
        }
    }

    private fun setupViewPager() {
        fragmentAdapter = UserProfileRecipesAdapter(this)
        println("UserProfileRecipesAdapter init")
        viewPager = vp_userprofile
        viewPager.isUserInputEnabled = false
        viewPager.adapter = fragmentAdapter
        viewPager.offscreenPageLimit = 2
        val tabLayout = tabLayout_userprofile
        TabLayoutMediator(tabLayout, viewPager) { tab, position ->
            tab.text =
                if (position == 0) {
                    resources.getString(R.string.title_recipes_profile)
                } else {
                    resources.getString(R.string.title_favourites_profile)
                }
        }.attach()
        //TODO fix
//        ViewPager2ViewHeightAnimator().viewPager2 = viewPager
    }

    private fun loadUser(username: String, updateCurrentUser: Boolean) {
        userViewModel.getUserByUsername(username, object : ResultCallback<User> {
            override fun onResult(value: User?) {
                if (value == null) {
                    loadUser(username, updateCurrentUser)
                } else {
                    if (updateCurrentUser) {
                        userViewModel.user.value = value
                        drawData(userViewModel.user.value!!)
                    } else {
                        userViewModel.selectedUser.value = value
                        drawData(userViewModel.selectedUser.value!!)
                    }
                }
            }

            override fun onFailure(value: User?) {
                loadUser(username, updateCurrentUser)
            }
        })
    }

    private fun focusOnRecipes() {
        nsv_userprofile.post {
            nsv_userprofile.smoothScrollTo(
                0, cardview_userprofile.bottom
            )
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        Log.d("UserProfileFragment", "onDestroy")
        userViewModel.selectedUsername.value = null
        userViewModel.selectedUser.value = null
    }

    override fun onPause() {
        super.onPause()
        Log.i("UserProfileFragment", "onPause")
    }

    override fun onResume() {
        super.onResume()
        Log.i("UserProfileFragment", "onResume")
    }

    override fun onStart() {
        super.onStart()
        Log.i("UserProfileFragment", "onStart")
    }

    override fun onBackPressedBase(): Boolean {
        return false
    }

    override fun scrollUpBase() {
        nsv_userprofile.smoothScrollTo(0, 0)
        appBarLayout_userprofile.setExpanded(true, false)
    }
}
